WATOLINK/bci-gaming
# Kevin character rewrite

## ClickUp ticket link

[Game - Improve Hitboxes](https://app.clickup.com/t/8685xn4m1)
[Improve and Rewrite Character scripts](https://app.clickup.com/t/8686d15jc)



## Implementation description

- Completely rewrote Player, Character, Enemy base scripts to use composition. 
  - To use, simply add a "Bandit" or "Player" script (make sure it's the new ones) and most of the required components will be added. 
  - Only the box collider component is not automatically added, in addition to the specific attributes of the components (i.e., sprite, animator, movement properties, sound effects, etc.)
  - Note that the Player uses two Box colliders so the sides have no friction (so the player doesn't stick to walls) but the feet still have friction
- Refactored new (and some old) scripts to use folders and namespaces.
- Added IPositionTrackable interface to allow other game objects (camera, enemies) to track an object's (i.e., the player's) position 
- Added IDamageable interface to test knockback on attack, I've found interfaces for health and damage to be pretty simple
  -  We can add onto IDamageable with health and death functions in the future (or remove it, I mainly added it to ensure knockback can works with my movement implementation)
  - It is currently implemented in Player, meaning only the player is "damageable" as of now
- Redid the camera script to have a script attached to the main camera (instead of a script attached to the player)
  - To use, add the PlayerCamera script onto the main camera, you can play with zoom in the camera object

(I tried to keep changes outside the new Assets/Scripts folder minimal to avoid merge conflicts and breaking existing scenes, but in the future most of the unorganized scripts will hopefully be deleted and/or moved into the Assets/Scripts folder)


## Steps to test
1. Open Assets/Demos/TestNewPlayer and press play
2. Ensure movement feels good, or can adjust the movement properties via the hierarchy: Entities/Player -> Player Movement (script)
3. Ensure no big movement bugs



## What should reviewers focus on?
Biggest thing is to make sure that movement is not broken. 

As for the code, look at the scripts inside of Assets/Scripts/... and check:
- Is the code neat and easy to understand? 
- Does it feel like it would be easily maintainable? 
- Would it be easy to add new features? 
- Should some things be done differently (code-wise or implementation-wise)?


## Checklist
